### DISCUSSION TITLE
Nitro Plugin access and extend h3 instance (server)

### DISCUSSION THREAD

**Discussion by masterjanic** [nuxt/nuxt#16663] [Questions]
I am currently using a nitro plugin to create a websocket server:

`export default defineNitroPlugin((nitro) => {
const wss = new WebSocketServer({ port: 1337 });
wss.on('connection', (ws) => { ... }
});`

This works perfectly fine, but the server needs to have a different port than the nuxt server.

My question is: How can I use the h3App instance for the Websocket Server?

I am using the 'ws' library. You could pass a server object like this: `new WebSocketServer({ server: httpServerInstance })`
But using the given nitro.h3App object throws an error `server.on is not defined` meaning this is not a httpserver instance.

Is there a possibility to access the instance?

**Comment by jbayona89** [comment]
I'm having the same issue, it would be nice to know how to access the h3 server instance...

**Comment by pi0** [comment]
Web sockets are not supported at the moment. We are working on it.

**Reply by vflam** [reply]
Has the situation evolved 1 year later? I'm having the same issue as OP and not finding a good solution. A workaround is to attach the websocket server inside a nuxt server middleware but it seems sketchy.

**Reply by dword-design** [reply]
It's the approach that I'm using right now too

```js
// model/io.js

const io = {}

export default io
```

```js
// server/middleware/socket-io.js

import { defineEventHandler } from '#imports'
import { Server } from 'socket.io'

import io from '@/model/io.js'

export default defineEventHandler(event => {
  if (!io.server && event.node.req.socket?.server) {
    io.server = new Server(event.node.req.socket.server)
  }
})
```

**Reply by marifcelik** [reply]
hi @dword-design is there an example for this ? i dont get it

**Reply by dword-design** [reply]
@marifcelik This is already the example. What are you missing? The middleware sets the server variable and then you can use it somewhere else in the backend to send messages.

**Reply by marifcelik** [reply]
there is not `server` under the `node.req.socket`. where is come from ?

**Reply by dword-design** [reply]
maybe there were changes. i'd need to have a look.

**Reply by muneebbug** [reply]
This solution worked perfectly for me. Thanks @dword-design

**Comment by lewebsimple** [comment]
In development mode, you can use the `listen` hook in a module to access the server:
```
import { defineNuxtModule } from "@nuxt/kit";

export default defineNuxtModule({
  setup(_options, nuxt) {
    nuxt.hook("listen", (server) => {
      const { address, port } = server.address() as any;
      console.log(`Server at ${address}:${port}`);
    });
  },
});
```

In the production build though, overriding the node preset to access the http server was pretty easy before `nitropack` came along (see my previous answer [here](https://github.com/nuxt/framework/discussions/3907#discussioncomment-2439203)), but it seems more difficult now (or at least I haven't seen any working example of this anywhere yet).

This is where I'm stuck right now:

Override `node-server` preset in `nuxt.config.ts`:
```
import { resolve } from "pathe";
import { defineNuxtConfig } from "nuxt";

export default defineNuxtConfig({
  nitro: {
    preset: resolve(__dirname, "server/index.ts"),
  },
});
```

Try to reproduce the [node-server preset](https://github.com/unjs/nitro/blob/main/src/runtime/entries/node-server.ts) in `server/index.ts`
```
import "node-fetch-native/polyfill"; // this was #internal/nitro/virtual/polyfill
import { Server as HttpServer } from "http";
import { Server as HttpsServer } from "https";
import destr from "destr";
import { nitroApp } from "nitropack/dist/runtime/app"; // this was "../app"
//import { useRuntimeConfig } from '#internal/nitro'; // replaced with Nuxt auto-imported useRuntimeConfig ... but it doesn't work!

const cert = process.env.NITRO_SSL_CERT;
const key = process.env.NITRO_SSL_KEY;

const server =
  cert && key ? new HttpsServer({ key, cert }, nitroApp.h3App.nodeHandler) : new HttpServer(nitroApp.h3App.nodeHandler);

const port = (destr(process.env.NITRO_PORT || process.env.PORT) || 3000) as number;
const hostname = process.env.NITRO_HOST || process.env.HOST || "0.0.0.0";

// @ts-expect-error ignore callback
server.listen(port, hostname, (err) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  const protocol = cert && key ? "https" : "http";
  console.log(`Listening on ${protocol}://${hostname}:${port}${useRuntimeConfig().app.baseURL}`);
});

if (process.env.DEBUG) {
  process.on("unhandledRejection", (err) => console.error("[nitro] [dev] [unhandledRejection]", err));
  process.on("uncaughtException", (err) => console.error("[nitro] [dev] [uncaughtException]", err));
} else {
  process.on("unhandledRejection", (err) => console.error("[nitro] [dev] [unhandledRejection] " + err));
  process.on("uncaughtException", (err) => console.error("[nitro] [dev] [uncaughtException] " + err));
}

export default {};
```

When building, I get `Cannot read properties of undefined (reading 'nitro')` at `node_modules/nitropack/dist/runtime/config.mjs:5:39` as the runtime config used is from Nuxt and not Nitro.

I'm praying @pi0 and @danielroe to shine their divine light upon us and help us reproduce the node-preset with the latest nitropack!

**Reply by lewebsimple** [reply]
A simple solution would be to expose the `listen` hook in nitropack ... is that a possibility ?
See https://github.com/unjs/nitro/discussions/270

**Reply by lewebsimple** [reply]
The new [node-server.ts](https://github.com/unjs/nitro/blob/main/src/runtime/entries/node-server.ts) entry has to be modified to import `nitroApp` from `#internal/nitro/app` instead of `../app`.

**Reply by paramono** [reply]
@lewebsimple have you managed to make it work on a recent version of Nuxt? (nuxt@3.0.0-rc.11)

I have problems with this import:
`import { nitroApp } from "nitropack/dist/runtime/app"; // this was "../app"`

It breaks with the following error:
```
ERROR  Cannot read properties of undefined (reading 'nitro')                                                                                                                                                                
  at node_modules/nitropack/dist/runtime/config.mjs:5:39
  at jiti (node_modules/jiti/dist/jiti.js:1:247497)
  at node_modules/nitropack/dist/runtime/app.mjs:7:15
  at jiti (node_modules/jiti/dist/jiti.js:1:247497)
  at server/index.ts:6:12
  at Object.jiti (node_modules/jiti/dist/jiti.js:1:247497)
  at resolveConfig (node_modules/nitropack/node_modules/c12/dist/index.mjs:220:21)
  at async extendConfig (node_modules/nitropack/node_modules/c12/dist/index.mjs:167:21)
  at async loadConfig (node_modules/nitropack/node_modules/c12/dist/index.mjs:126:5)
  at async loadOptions (node_modules/nitropack/dist/shared/nitro.e0bd1b3d.mjs:2229:30)
```

This seems to happen because `_runtimeConfig` is empty and hence its `nitro` property cannot be accessed:
https://github.com/unjs/nitro/blob/635f5e43187eb37bf166c4b3ae088d4c594de945/src/runtime/config.ts#L5-L8

Should I do something to make sure this config is not empty?

I found a few alternative approaches but none of them work for me.

For example, here
https://github.com/stafyniaksacha/nuxt-poc-websockets/blob/c758251c30815d02b04fde2a1722c80f3433f67d/start.mjs#L3

I changed the import to:
`import { a as useRuntimeConfig, u as useNitroApp } from '../.output/server/chunks/nitro/node-server.mjs'`

But this module fails to `import 'node-fetch-native/polyfill'` (attempts to load an adsent .cjs file instead of the existing .mjs one)   

Any way to access a nitroApp from a custom nitro preset?

**Reply by paramono** [reply]
I've managed to make it work by using this approach:
https://github.com/nuxt/framework/discussions/3907#discussioncomment-2439203

Basically, In nuxt.config.ts, I point preset setting to a module with the following contents:

```
import type { NitroPreset } from "nitropack";

export default {
  entry: "server/entry",
  externals: true,
  serveStatic: true,
} as NitroPreset;
```

And in `server/entry.ts` you  reproduce a nitro preset with all the necessary modifications, similar to the one @lewebsimple
 posted above
